pkg/cache/v3/delta.go

```diff
@@ -16,73 +16,292 @@ package cache
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"sync/atomic"
 
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
 	"github.com/envoyproxy/go-control-plane/pkg/cache/types"
-	"github.com/envoyproxy/go-control-plane/pkg/server/stream/v3"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
 )
 
+type resourceWithNameAndVersion interface {
+	getResource() types.Resource
+	getName() string
+	getVersion() string
+}
+
+func getResourceNames(resources []resourceWithNameAndVersion) []string {
+	names := make([]string, 0, len(resources))
+	for _, res := range resources {
+		names = append(names, res.getName())
+	}
+	return names
+}
+
+var _ DeltaResponse = &rawDeltaResponse{}
+
+// RawDeltaResponse is a pre-serialized xDS response that utilizes the delta discovery request/response objects.
+type rawDeltaResponse struct {
```

Comment:
Made private. This allows us to evolve Cache implementations with less impact on users